docs/features/tuples.md

```diff
@@ -0,0 +1,227 @@
+var t = ((int foo, int bar))o;		 // unboxing conversion
+```
+
+Target typing
+--------------
+
+A tuple literal is "target typed" when used in a context specifying a tuple type. What that means is that the tuple literal has a "conversion from expression" to any tuple type, as long as the element expressions of the tuple literal have an implicit conversion to the element types of the tuple type.
+
+``` c#
+(string name, byte age) t = (null, 5); // Ok: the expressions null and 5 convert to string and byte
+```
+
+In cases where the tuple literal is not part of a conversion, a tuple is used by its "natural type", which means a tuple type where the element types are the types of the constituent expressions. Since not all expressions have types, not all tuple literals have a natural type either:
+
+``` c#
+var t = ("John", 5);  						//   Ok: the type of t is (string, int)
+var t = (null, 5);    						//   Error: tuple expression doesn't have a type because null does not have a type
+((1,2, null), 5).ToString();    	    	//   Error: tuple expression doesn't have a type
+
+ImmutableArray.Create((()=>1, 1));        	//   Error: tuple expression doesn't have a type because lambda does not have a type
+ImmutableArray.Create((Func<int>)()=>1, 1); //   ok
+```
+
+A tuple literal may include names, in which case they become part of the natural type:
+
+``` c#
+var t = (name: "John", age: 5); // The type of t is (string name, int age)
+t.age++;						// t has field named age.
+```
+
+A successful conversion from tuple expression to tuple type is classified as _ImplictTuple_ conversion, unless tuple's natural type matches the target type exactly, in such case it is an _Identity_ conversion.
+
+```C#
+void foo((int x, int y) arg){...};
+void foo((object x, object y) arg){...};
+
+foo((1, 2));   			// first overload is used. Identity conversion is better than implicit conversion. 
+foo(("hi", "hello"));   // second overload is used. Implicit tuple conversion is better than no conversion. 
+```
+
+Target typing will "see through" nullable target types. A successfull conversion from tuple expression to a nullable tuple type is classified as _ImplicitNullable_ conversion.
+
+```C#
+((int x, int y, int z)?, int t)? SpaceTime()
+{
+	return ((1,2,3), 7); 	// valid, implicit nullable conversion
+}
+``` 
+
+Overload resolution and tuples with no natural types.
+--------------
+
+A situation may arise during overload resolution where multiple equally qualified candidates could be available due to a tuple argument without natural type being implicitly convertible to the corresponding parameters.
+
+In such situations overload resolution employs _exact match_ rule where arguments without natural types are recursively matched to the types of the corresponding parameters in terms of constituent structural elements.
+
+Tuple expressions are no exception from this rule and the _exact match_ rule is based on the natural types of the constituent tuple arguments.
```

Comment:
When an expression is a tuple with natural type and it matches the target tuple type it implies *exact match*. 
We have a choice here - to require recursion in all tuples or only into those with no natural type. The result is the same.

Current implementation does not recurse when natural type is known. Perhaps it is just an optimization. It is a significant optimization though since having natural type is a very common case.